File: src/js/fable-library/FSharp.Collections.fs
namespace FSharp.Core

open System.Collections.Generic

module HashIdentity =
    let FromFunctions hash eq : IEqualityComparer<'T> =
        { new IEqualityComparer<'T> with
            member __.GetHashCode(x) = hash x
            member __.Equals(x,y) = eq x y }
    let Structural<'T when 'T : equality> : IEqualityComparer<'T> =
        FromFunctions LanguagePrimitives.GenericHash LanguagePrimitives.GenericEquality
    let Reference<'T when 'T : not struct > : IEqualityComparer<'T> =
        FromFunctions LanguagePrimitives.PhysicalHash LanguagePrimitives.PhysicalEquality

module ComparisonIdentity = 
    let FromFunction comparer : IComparer<'T> =
        { new IComparer<'T> with
            member __.Compare(x,y) = comparer x y }
    let Structural<'T when 'T : comparison> : IComparer<'T> =
        FromFunction LanguagePrimitives.GenericComparison

File: src/fable-compiler-js/fable-library/FSharp.Collections.js
import { compare, physicalHash, equals, structuralHash } from "./Util.js";
export function HashIdentity_FromFunctions(hash, eq) {
    return {
        Equals(x, y) {
            return eq(x, y);
        },
        GetHashCode(x_1) {
            return hash(x_1);
        },
    };
}
export function HashIdentity_Structural() {
    return HashIdentity_FromFunctions(structuralHash, equals);
}
export function HashIdentity_Reference() {
    return HashIdentity_FromFunctions(physicalHash, (e, e_1) => (e === e_1));
}
export function ComparisonIdentity_FromFunction(comparer) {
    return {
        Compare(x, y) {
            return comparer(x, y);
        },
    };
}
export function ComparisonIdentity_Structural() {
    return ComparisonIdentity_FromFunction(compare);
}
